// maybeasy/script/src/Just.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.just = exports.Just = void 0;
const Maybe_js_1 = __importDefault(require("./Maybe.js"));
class Just extends Maybe_js_1.default {
    constructor(value) {
        super();
        Object.defineProperty(this, "value", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: value
        });
    }
    getOrElse(_fn) {
        return this.value;
    }
    getOrElseValue(_defaultValue) {
        return this.value;
    }
    map(fn) {
        return new Just(fn(this.value));
    }
    andThen(fn) {
        return fn(this.value);
    }
    orElse(_fn) {
        return this;
    }
    cata(matcher) {
        return matcher.Just(this.value);
    }
    assign(k, other) {
        const maybe = typeof other === 'function' ? other(this.value) : other;
        return maybe.map(b => ({
            ...Object(this.value),
            [k.toString()]: b
        }));
    }
    do(fn) {
        fn(this.value);
        return new Just(this.value);
    }
    elseDo(_fn) {
        return this;
    }
    isJust() {
        return true;
    }
    isNothing() {
        return false;
    }
}
exports.Just = Just;
const just = (value) => new Just(value);
exports.just = just;
exports.default = Just;
