// maybeasy/script/src/Nothing.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nothing = exports.Nothing = void 0;
const Maybe_js_1 = __importDefault(require("./Maybe.js"));
class Nothing extends Maybe_js_1.default {
    getOrElse(fn) {
        return fn();
    }
    getOrElseValue(defaultValue) {
        return defaultValue;
    }
    map(_fn) {
        return new Nothing();
    }
    andThen(_fn) {
        return new Nothing();
    }
    orElse(fn) {
        return fn();
    }
    cata(matcher) {
        return matcher.Nothing();
    }
    assign(_k, _other) {
        return new Nothing();
    }
    do(_fn) {
        return new Nothing();
    }
    elseDo(fn) {
        fn();
        return this;
    }
    isJust() {
        return false;
    }
    isNothing() {
        return true;
    }
}
exports.Nothing = Nothing;
const nothing = () => new Nothing();
exports.nothing = nothing;
exports.default = Nothing;
